File: apkEditor/src/main/java/com/mcal/apkeditor/ApkComposeFailAdapter.java
package com.mcal.apkeditor;

import android.content.Intent;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.BaseAdapter;

import androidx.appcompat.widget.AppCompatButton;
import androidx.appcompat.widget.AppCompatTextView;

import com.mcal.apkeditor.activities.ApkComposeActivity;
import com.mcal.apkeditor.editor.TextEditor;
import com.mcal.common.utils.ActivityUtils;

import java.io.File;
import java.lang.ref.WeakReference;
import java.util.ArrayList;
import java.util.List;

public class ApkComposeFailAdapter extends BaseAdapter {

    private final WeakReference<ApkComposeActivity> activityRef;
    private final List<String> lines = new ArrayList<>();

    public ApkComposeFailAdapter(ApkComposeActivity activity, String errMessage) {
        this.activityRef = new WeakReference<>(activity);
        updateMessage(errMessage);
    }

    public void updateMessage(String errMessage) {
        lines.clear();

        if (errMessage != null) {
            String[] arr = errMessage.split("\\r?\\n");
            for (String line : arr) {
                // Make sure not empty
                if (!"".equals(line)) {
                    lines.add(line);
                }
            }
        }
    }

    @Override
    public int getCount() {
        return lines.size();
    }

    @Override
    public Object getItem(int position) {
        return lines.get(position);
    }

    @Override
    public long getItemId(int position) {
        return position;
    }

    @Override
    public View getView(int position, View convertView, ViewGroup parent) {
        ViewHolder holder;
        if (convertView == null) {
            convertView = LayoutInflater.from(activityRef.get()).inflate(R.layout.item_failed_message, null);
            holder = new ViewHolder();
            holder.messageTv = convertView.findViewById(R.id.message);
            holder.viewBtn = convertView.findViewById(R.id.btn_view);
            convertView.setTag(holder);
        } else {
            holder = (ViewHolder) convertView.getTag();
        }

        // Set on click listener
        String strLine = lines.get(position);
        boolean viewable = false;
        int commaPos = strLine.indexOf(':');
        if (commaPos != -1) {
            final String filePath = strLine.substring(0, commaPos);
            if (new File(filePath).exists()) {
                viewable = true;
                holder.viewBtn.setVisibility(View.VISIBLE);

                // Try to get the line number
                int lineNum = -1;
                int nextCommaPos = strLine.indexOf(':', commaPos + 1);
                if (nextCommaPos != -1) {
                    try {
                        lineNum = Integer.parseInt(strLine.substring(commaPos + 1, nextCommaPos));
                    } catch (Exception ignored) {
                    }
                }

                final int lineIndex = lineNum;
                holder.viewBtn.setOnClickListener(v -> {
                    ApkComposeActivity activity = activityRef.get();
                    Intent intent = TextEditor.getSoraEditor(activity, filePath, activity.srcApkPath);
                    if (lineIndex > 0) {
                        ActivityUtils.attachParam(intent, "startLine", "" + lineIndex);
                    }
                    activity.startActivity(intent);
                });
            }
        }
        if (!viewable) { // Cannot view(open the file) for this line
            holder.viewBtn.setVisibility(View.INVISIBLE);
        }

        holder.messageTv.setText(lines.get(position));

        return convertView;
    }

    private static class ViewHolder {
        AppCompatTextView messageTv;
        AppCompatButton viewBtn;
    }
}

File: apkEditor/src/main/java/com/mcal/apkeditor/adapters/ApkComposeFailAdapter.kt
package com.mcal.apkeditor.adapters

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Button
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import com.mcal.apkeditor.R
import com.mcal.apkeditor.activities.ApkComposeActivity
import com.mcal.common.utils.ScopedStorage
import com.mcal.editor.TextEditor.getSoraEditor
import java.io.File
import java.util.regex.Pattern

class ApkComposeFailAdapter(
    private val activity: ApkComposeActivity, errMessage: String?
) : RecyclerView.Adapter<ApkComposeFailAdapter.ApkComposeFailViewHolder>() {
    private val lines: MutableList<String> = ArrayList()

    init {
        updateMessage(errMessage)
    }

    private fun updateMessage(errMessage: String?) {
        if (errMessage != null) {
            lines.add(errMessage)
        }
    }

    override fun getItemCount(): Int {
        return lines.size
    }

    fun getItem(position: Int): Any {
        return lines[position]
    }

    override fun getItemId(position: Int): Long {
        return position.toLong()
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ApkComposeFailViewHolder {
        val itemView = LayoutInflater.from(parent.context).inflate(R.layout.item_failed_message, parent, false)
        return ApkComposeFailViewHolder(itemView)
    }

    override fun onBindViewHolder(holder: ApkComposeFailViewHolder, position: Int) {
        val strLine = lines[position]
        if (strLine.isEmpty()) {
            return
        }

        var filePath: String? = null
        var lineIndex = 0

        /**
         * Пример ошибки в smali: Source: data/data/com.mcal.apkeditor.pro/files/ApkProtector/smali/a/a.smali;Line: 15;Column: 0
         * Пример ошибки в xml: Source: data/data/com.mcal.apkeditor.pro/files/ApkProtector/AndroidManifest.xml;Line: 15;Message: not well-formed (invalid token)
         */
        val pathMatcher = Pattern.compile("(Source: )(.+)(;)(Line: )(\\d+)(;)((Message: )|(Column: ))(.*)").matcher(strLine)
        if (pathMatcher.find()) {
            filePath = pathMatcher.group(2)?.takeIf { File(it).exists() }?.also {
                /**
                 * Получение пути к файлу с ошибкой
                 */
                holder.pathView.text = buildString {
                    append(activity.getString(R.string.error_path))
                    append(it.replace(ScopedStorage.filesDir.path, ""))
                }
            }
            pathMatcher.group(5)?.let {
                /**
                 * Получение номера строки ошибки
                 */
                holder.lineView.text = buildString {
                    append(activity.getString(R.string.error_line))
                    append(it)
                }
                lineIndex = it.toInt() - 1
            }
            val messageKey = pathMatcher.group(7)
            if (messageKey != null) {
                if (messageKey.startsWith("Message")) {
                    /**
                     * Получение сообщения ошибки - только для xml
                     */
                    pathMatcher.group(10)?.let { value ->
                        holder.messageView.text = buildString {
                            append(activity.getString(R.string.error_message))
                            append(value)
                        }
                    }
                } else if (messageKey.startsWith("Column")) {
                    /**
                     * Получение столбца ошибки - только для smali
                     */
                    pathMatcher.group(10)?.let { value ->
                        holder.messageView.text = buildString {
                            append(activity.getString(R.string.error_column))
                            append(value)
                        }
                    }
                }
            } else {
                /**
                 * Иначе отображаем полную ошибку, как есть
                 */
                holder.messageView.text = buildString {
                    append(activity.getString(R.string.error_message))
                    append(strLine)
                }
            }
        }
        filePath?.let { path ->
            /**
             * Если путь не пустой предлагаем открыть в редакторе
             */
            holder.editor.setOnClickListener {
                val intent = getSoraEditor(activity, path, activity.srcApkPath, lineIndex, null)
                activity.startActivity(intent)
            }
        } ?: run {
            holder.editor.visibility = View.GONE
        }
    }

    class ApkComposeFailViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        val pathView: TextView = itemView.findViewById(R.id.path)
        val lineView: TextView = itemView.findViewById(R.id.line)
        val messageView: TextView = itemView.findViewById(R.id.message)
        val editor: Button = itemView.findViewById(R.id.editor)
    }
}